apps/accounts/views.py

```diff
@@ -72,72 +99,100 @@ def login_user(request, user, identity):
     login(request, user)
 
 
-def with_user(fn):
-    @functools.wraps(fn)
-    def inner(self, request):
-        data = request.GET if request.method == 'GET' else request.DATA
-        state_parts = data.get('state', '').split(':', 1)
-        state = state_parts[0]
-        if 'code' not in data:
-            log.info('No code provided.')
-            return Response({'error': 'No code provided.'}, status=422)
-        elif (not request.session.get('fxa_state') or
-                request.session['fxa_state'] != state):
-            log.info('State mismatch. URL: {url} Session: {session}'.format(
-                url=data.get('state'),
-                session=request.session.get('fxa_state'),
-            ))
-            return Response({'error': 'State mismatch.'}, status=400)
+def render_error(request, error, next_path=None, format=None):
+    if format == 'json':
+        status = ERROR_STATUSES[error]
+        return Response({'error': error}, status=status)
+    else:
+        if not is_safe_url(next_path):
+            next_path = None
+        return HttpResponseRedirect(
+            urlparams(reverse('users.login'), error=error, to=next_path))
+
 
+def parse_next_path(state_parts):
+    next_path = None
+    if len(state_parts) == 2:
+        # The = signs will be stripped off so we need to add them back
+        # but it only cares if there are too few so add 4 of them.
+        encoded_path = state_parts[1] + '===='
```

Comment:
ok, seems a little fragile to tack on 4 equal signs but as long as there is test coverage I'm fine with it.